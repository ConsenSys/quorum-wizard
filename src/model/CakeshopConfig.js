import {
  isDocker,
  isTessera,
} from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function generateCakeshopConfig(config) {
  const nodes = []

  config.nodes.forEach((node, i) => {
    const nodeData = {
      name: `node${i + 1}`,
      rpcUrl: isDocker(config.network.deployment)
        // docker-compose creates dns entries for the container's ip using container name (nodeX)
        ? `http://node${i + 1}:${config.containerPorts.quorum.rpcPort}`
        : `http://localhost:${node.quorum.rpcPort}`,
      transactionManagerUrl: createTransactionManagerUrl(config, node, i),
    }
    nodes.push(nodeData)
  })
  return nodes
}

function createTransactionManagerUrl(config, node, i) {
  if (isTessera(config.network.transactionManager)) {
    return isDocker(config.network.deployment)
      // docker-compose creates dns entries for the container's ip using container name (txmanagerX)
      ? `http://txmanager${i + 1}:${config.containerPorts.tm.thirdPartyPort}`
      : `http://localhost:${node.tm.thirdPartyPort}`
  }
  return ''
}
