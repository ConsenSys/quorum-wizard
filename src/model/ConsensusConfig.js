import {
  writeFile,
  writeJsonFile,
  combineFiles,
  readFileToString
} from '../utils/fileUtils'
import { join } from 'path'
import { executeSync } from '../utils/execUtils'
const Web3 = require('web3');

export function generateConsensusConfig(configDir, consensus, nodes) {
  return consensus == `raft` ?  generateRaft(configDir, nodes) :
  generateIstanbul(configDir, nodes)
}

function generateRaft(configDir, nodes) {
  writeJsonFile(configDir, `raft-genesis.json`, generateRaftConfig(nodes, configDir))
}

function generateIstanbul(configDir, nodes) {
  writeJsonFile(configDir, `istanbul-genesis.json`, generateIstanbulConfig(nodes, configDir))
}

export function generateAccounts(nodes, configDir) {
  var numNodes = nodes.length
  var raftJsonString = `{`
  for (let i = 1; i < parseInt(numNodes, 10); i++) {
    let keyDir = join(configDir, `key${i}`)
    let keyString = readFileToString(join(keyDir, 'key'))

    raftJsonString += `\"${JSON.parse(keyString).address}\":{\"balance\":\"1000000000000000000000000000\"},`
  }
  let keyDir = join(configDir, `key${numNodes}`)
  let keyString = readFileToString(join(keyDir, 'key'))

  raftJsonString += `\"${JSON.parse(keyString).address}\":{\"balance\":\"1000000000000000000000000000\"}}`
  return JSON.parse(raftJsonString)
}

function generateRaftConfig(numNodes, configDir) {
  const alloc = generateAccounts(numNodes, configDir)
  return {
    alloc: alloc,
    coinbase: `0x0000000000000000000000000000000000000000`,
    config: {
      homesteadBlock: 0,
      byzantiumBlock: 0,
      constantinopleBlock: 0,
      chainId: 10,
      eip150Block: 0,
      eip155Block: 0,
      eip150Hash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
      eip158Block: 0,
      maxCodeSize: 35,
      isQuorum: true,
    },
    difficulty: `0x0`,
    extraData: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    gasLimit: `0xE0000000`,
    mixhash: `0x00000000000000000000000000000000000000647572616c65787365646c6578`,
    nonce: `0x0`,
    parentHash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    timestamp: `0x00`
  }
}

function generateExtraData(nodes, configDir) {
  var validatorConfigString = `vanity = \"0x00\"\nvalidators=[`
  var numNodes = nodes.length
  for (let i = 1; i < parseInt(numNodes, 10); i++) {
      let keyDir = join(configDir, `key${i}`)
      let nodekey = nodekeyToAccount(readFileToString(join(keyDir, 'nodekey')))
      validatorConfigString += `\"${nodekey}\",`
  }
  let keyDir = join(configDir, `key${numNodes}`)
  let nodekey = nodekeyToAccount(readFileToString(join(keyDir, 'nodekey')))
  validatorConfigString += `\"${nodekey}\"]`

  var istanbulConfigFile = join(configDir, 'istanbul.toml')
  writeFile(istanbulConfigFile, validatorConfigString, false)

  var extraDataCmd = `cd ${configDir} && istanbul extra encode --config ${istanbulConfigFile} | awk '{print $4}' > extraData`
  executeSync(extraDataCmd, function(err, stdout, stderr) {
    if (e instanceof Error) {
      console.error(e)
      throw e
    }
  })
  return readFileToString(join(configDir, 'extraData'))
}

function nodekeyToAccount(nodekey) {
  const web3 = new Web3();
  var acctObj = web3.eth.accounts.privateKeyToAccount(nodekey);
  return acctObj.address
}

function generateIstanbulConfig(nodes, configDir) {
  const alloc = generateAccounts(nodes, configDir)
  const extraData = generateExtraData(nodes, configDir)
  return {
    alloc: alloc,
    coinbase: `0x0000000000000000000000000000000000000000`,
    config: {
      homesteadBlock: 0,
      byzantiumBlock: 0,
      constantinopleBlock: 0,
      chainId: 10,
      eip150Block: 0,
      eip155Block: 0,
      eip150Hash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
      eip158Block: 0,
      maxCodeSize: 35,
      isQuorum: true,
      istanbul: {
        epoch: 30000,
        policy: 0,
        ceil2Nby3Block: 0,
      }
    },
    difficulty: `0x1`,
    extraData: extraData,
    gasLimit: `0xE0000000`,
    mixhash: `0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365`,
    nonce: `0x0`,
    parentHash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    timestamp: `0x00`
  }

}
