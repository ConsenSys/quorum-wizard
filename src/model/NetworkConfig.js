import { isJava11Plus } from '../utils/execUtils'
import {
  LATEST_CAKESHOP,
  LATEST_CAKESHOP_J8,
  LATEST_QUORUM,
  LATEST_TESSERA,
  LATEST_TESSERA_J8,
} from '../generators/download'
import {
  getDockerSubnet,
  cidrhost,
} from '../utils/subnetUtils'
import { getOutputPath } from '../utils/fileUtils'

export function createConfigFromAnswers(answers) {
  const {
    name,
    networkPath = getOutputPath(),
    numberNodes = 3,
    consensus = 'raft',
    quorumVersion = LATEST_QUORUM,
    transactionManager = isJava11Plus() ? LATEST_TESSERA : LATEST_TESSERA_J8,
    deployment = 'bash',
    cakeshop = isJava11Plus() ? LATEST_CAKESHOP : LATEST_CAKESHOP_J8,
    generateKeys = false,
    networkId = '10',
    genesisLocation = 'none',
    customizePorts = false,
    nodes = [],
    cakeshopPort = '8999',
    remoteDebug = false,
    containerPorts = undefined,
  } = answers
  const networkFolder = name
    || defaultNetworkName(numberNodes, consensus, transactionManager, deployment)
  const dockerSubnet = (isDocker(deployment) && containerPorts !== undefined) ? containerPorts.dockerSubnet : ''
  return {
    network: {
      name: networkFolder,
      networkPath,
      verbosity: 5,
      consensus,
      quorumVersion,
      transactionManager,
      permissioned: true,
      genesisFile: genesisLocation,
      generateKeys,
      configDir: `network/${networkFolder}/resources`,
      deployment,
      cakeshop,
      networkId,
      customizePorts,
      cakeshopPort,
      remoteDebug,
    },
    nodes: (customizePorts && nodes.length > 0) ? nodes : generateNodeConfigs(
      numberNodes,
      transactionManager,
      deployment,
      cakeshop,
      dockerSubnet,
    ),
    containerPorts,
  }
}

export function defaultNetworkName(numberNodes, consensus, transactionManager, deployment) {
  const transactionManagerName = !isTessera(transactionManager)
    ? ''
    : 'tessera-'
  return `${numberNodes}-nodes-${consensus}-${transactionManagerName}${deployment}`
}

export function generateNodeConfigs(
  numberNodes,
  transactionManager,
  deployment,
  cakeshop,
  dockerSubnet,
) {
  const devP2pPort = 21000
  const rpcPort = 22000
  const wsPort = 23000
  const graphQlPort = 24000
  const raftPort = 50401
  const thirdPartyPort = 9081
  const p2pPort = 9001
  const nodes = []

  for (let i = 0; i < parseInt(numberNodes, 10); i += 1) {
    const node = {
      quorum: {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 10) : '127.0.0.1',
        devP2pPort: devP2pPort + i,
        rpcPort: rpcPort + i,
        wsPort: wsPort + i,
        raftPort: raftPort + i,
        graphQlPort: graphQlPort + i,
      },
    }
    if (isTessera(transactionManager)) {
      node.tm = {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 100) : '127.0.0.1',
        thirdPartyPort: thirdPartyPort + i,
        p2pPort: p2pPort + i,
      }
    }
    nodes.push(node)
  }
  return nodes
}

export function getContainerPorts(deployment) {
  const dockerSubnet = isDocker(deployment) ? getDockerSubnet() : ''
  return {
    dockerSubnet,
    quorum: {
      rpcPort: 8545,
      p2pPort: 21000,
      raftPort: 50400,
      wsPort: 8645,
      graphQlPort: 8547,
    },
    tm: {
      p2pPort: 9000,
      thirdPartyPort: 9080,
    },
  }
}

export function isTessera(tessera) {
  return tessera !== 'none'
}

export function isDocker(deployment) {
  return deployment === 'docker-compose'
}

export function isBash(deployment) {
  return deployment === 'bash'
}

export function isKubernetes(deployment) {
  return deployment === 'kubernetes'
}

export function isIstanbul(consensus) {
  return consensus === 'istanbul'
}

export function isRaft(consensus) {
  return consensus === 'raft'
}

export function isCakeshop(cakeshop) {
  return cakeshop !== 'none'
}
