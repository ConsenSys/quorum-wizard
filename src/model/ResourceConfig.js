import { isTessera } from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function buildKubernetesResource(config) {
  return [
    buildGeneralDetails(config),
    buildQuorumDetails(config),
    buildTesseraDetails(config),
    buildGethDetails(config),
  ].join('')
}

function buildGeneralDetails(config) {
  const nodeNum = config.nodes.length
  return `#namespace:
#  name: quorum-test
sep_deployment_files: true
generate_keys: ${config.network.generateKeys}
nodes:
  number: ${nodeNum}
service:
  type: NodePort`
}

function buildQuorumDetails(config) {
  return `
quorum:
  consensus: ${config.network.consensus}
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: ${config.nodes[0].quorum.raftPort}
    Quorum_Version: ${config.network.quorumVersion}
  storage:
    Type: PVC
    Capacity: 200Mi`
}

function buildGethDetails(config) {
  return `
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 30303
  network:
    id: ${config.network.networkId}
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\"*\\"`
}

function buildTesseraDetails(config) {
  const hasTessera = isTessera(config.network.transactionManager)
  return hasTessera
    ? `
  tm:
    Name: tessera
    Tm_Version: ${config.network.transactionManager}
    Port: ${config.nodes[0].tm.thirdPartyPort}
    Tessera_Config_Dir: out/config`
    : ''
}
