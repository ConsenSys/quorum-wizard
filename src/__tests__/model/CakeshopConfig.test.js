import { createQuickstartConfig } from '../../model/NetworkConfig'
import { generateCakeshopConfig } from '../../model/CakeshopConfig'

test('creates 3nodes raft dockerFile tessera cakeshop', () => {
  const config = createQuickstartConfig('3', 'raft', 'tessera', 'docker-compose', 'yes')
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes istanbul bash tessera cakeshop', () => {
  const config = createQuickstartConfig('3', 'istanbul', 'tessera', 'bash', 'yes')
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes raft dockerFile no tessera cakeshop', () => {
  const config = createQuickstartConfig('3', 'raft', 'none', 'docker-compose', 'yes')
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes istanbul bash no tessera cakeshop', () => {
  const config = createQuickstartConfig('3', 'istanbul', 'none', 'bash', 'yes')
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})
