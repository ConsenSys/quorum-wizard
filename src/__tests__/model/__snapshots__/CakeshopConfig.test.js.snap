// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates 3nodes istanbul bash no tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://localhost:22000",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://localhost:22001",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://localhost:22002",
    "transactionManagerUrl": "",
  },
]
`;

exports[`creates 3nodes istanbul bash tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://localhost:22000",
    "transactionManagerUrl": "http://localhost:9081/partyinfo/keys",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://localhost:22001",
    "transactionManagerUrl": "http://localhost:9082/partyinfo/keys",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://localhost:22002",
    "transactionManagerUrl": "http://localhost:9083/partyinfo/keys",
  },
]
`;

exports[`creates 3nodes raft dockerFile no tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://host.docker.internal:22000",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://host.docker.internal:22001",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://host.docker.internal:22002",
    "transactionManagerUrl": "",
  },
]
`;

exports[`creates 3nodes raft dockerFile tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://host.docker.internal:22000",
    "transactionManagerUrl": "http://host.docker.internal:9081/partyinfo/keys",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://host.docker.internal:22001",
    "transactionManagerUrl": "http://host.docker.internal:9082/partyinfo/keys",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://host.docker.internal:22002",
    "transactionManagerUrl": "http://host.docker.internal:9083/partyinfo/keys",
  },
]
`;
